proxy/server/server.go

```diff
@@ -270,30 +281,91 @@ func (s *T) newTLSCfg(hostCfgs map[engine.HostKey]engine.Host) (*tls.Config, err
 
 	defaultHostName := ""
 	pairs := map[string]tls.Certificate{}
+	getCertFuncs := map[string]getCertificateFunc{}
+
 	for _, hostCfg := range hostCfgs {
 		if hostCfg.Settings.Default {
 			defaultHostName = hostCfg.Name
 		}
-		c := hostCfg.Settings.KeyPair
-		if c == nil {
-			continue
-		}
-		keyPair, err := tls.X509KeyPair(c.Cert, c.Key)
-		if err != nil {
-			return nil, err
-		}
-		if hostCfg.Settings.OCSP.Enabled {
-			log.Infof("%v OCSP is enabled for %v, resolvers: %v", s, hostCfg, hostCfg.Settings.OCSP.Responders)
-			r, err := s.stapler.StapleHost(&hostCfg)
+		if hostCfg.Settings.KeyPair != nil {
+			c := hostCfg.Settings.KeyPair
+			if hostCfg.Settings.AutoCert != nil {
+				log.Warningf("Host %v has a KeyPair and AutoCert enabled. KeyPair takes precedence. Autocert generation disabled.", hostCfg.Name)
+			}
+			keyPair, err := tls.X509KeyPair(c.Cert, c.Key)
+			if err != nil {
+				return nil, err
+			}
+			if hostCfg.Settings.OCSP.Enabled {
+				log.Infof("%v OCSP is enabled for %v, resolvers: %v", s, hostCfg, hostCfg.Settings.OCSP.Responders)
+				r, err := s.stapler.StapleHost(&hostCfg)
+				if err != nil {
+					log.Warningf("%v failed to staple %v, error %v", s, hostCfg, err)
+				} else if r.Response.Status == ocsp.Good || r.Response.Status == ocsp.Revoked {
+					keyPair.OCSPStaple = r.Staple
+				} else {
+					log.Warningf("%s got undefined status from OCSP responder: %v", s, r.Response.Status)
+				}
+			}
+			pairs[hostCfg.Name] = keyPair
+		} else if hostCfg.Settings.AutoCert != nil {
+			ac := hostCfg.Settings.AutoCert
+			if s.autoCertMgrs == nil {
+				s.autoCertMgrs = make([]*autocert.Manager, 0, 0)
+			}
+
+			// Each host gets its own Autocert Manager - this allows individual
+			// certs to use different autocert authorities, as well as auth keys
+			autoCertMgr := &autocert.Manager{
+				Prompt:     autocert.AcceptTOS,
+				Cache:      s.autoCertCache,
+				HostPolicy: autocert.HostWhitelist(hostCfg.Name),
+				Email:      ac.Email,
+			}
+
+			if ac.RenewBeforeSeconds > 0 {
+				autoCertMgr.RenewBefore = time.Duration(ac.RenewBeforeSeconds) * time.Second
+			}
+
+			if ac.Key != "" || ac.DirectoryURL != "" {
+				autoCertMgr.Client = &acme.Client{
+					DirectoryURL: ac.DirectoryURL,
+				}
+
+				if ac.Key != "" {
+					block, _ := pem.Decode([]byte(ac.Key))
+					if block == nil {
+						log.Errorf("Autocert Key PEM Block for Host %s is invalid.", hostCfg.Name)
```

Comment:
Why do not return an error here and in the following checks?

Reply:
Thanks for catching that. I'll just "continue" the loop at the failure.